2016/proposals/rai-vivek.md

```diff
@@ -0,0 +1,360 @@
+* Version number of the found (installed) softwares
+* Any error that may have occurred in the script itself
+* Process information
+* Timestamp
+* (Optional) User email
+
+
+The current installation-test script implementation already collects the list of
+*successful* and *failed* checks along with version information and error
+message as available.
+
+### Server
+
+The server will collect the diagnostic report submitted by installation-scripts
+when they are run on the user machine.
+
+A simple design of the setup is discussed below:
+
+#### API
+
+The following steps are involved in the process:
+
+* Installation-script sends a `POST` request to the server endpoint (for example, https://installation.software-carpentry.org/report).
+  > Using Flask web-development framework for Python. A robust, flexible
+  > choice that is easy to develop and maintain. Scalability is not an issue for
+  > our use case.
+
+* Server verifies the data, open database connection, updates the record.
+  > As suggested in the ideas sections, SQLite will be a low-barrier, easier to
+  > maintain database, appropriate for our needs.
+
+* Send a status response back to the client. The response can contain feedback,
+  suggestions or perhaps link to some survey.
+
+* Ability to export diagnostic data into JSON or CSV format.
+  > It’ll allow us to separate the user interface from backend creating a modular
+  > setup.
+
+#### Database
+
+Based on the current nature of collectable information, write-volume and
+functionality required SQLite seems to be a fit choice. In this regard, the
+database schema is an important component that determines how we are going to
+store the collected data in the database.
+
+The following illustrates a typical schema that might be used:
+
+```
+TABLE machine_info {
+  id PRIMARY KEY AUTO,
+  system VARCHAR,
+  hostname VARCHAR,
+  release VARCHAR,
+  version VARCHAR,
+  machine VARCHAR,
+  processor VARCHAR,
+  email VARCHAR,
+  time VARCHAR,
+  exit_status INT
+}
+
+
+TABLE software_version {
+  id PRIMARY KEY AUTO,
+  software VARCHAR,
+  installed BOOL,
+  version VARCHAR,
+  log VARCHAR
+}
+```
+
+Since we are obtaining the platform information using Python `platform` module
+which is same across all platforms, the same database schema can accommodate all
+systems.
+
+A typical example of the database entry then may look like:
+
+```
+(TABLE machine_info)
+
+id | system  | hostname    | release             | version                                               | machine  | processor | email       | time                      | exit_status
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+1  | ‘Linux’ | ‘ghostbook’ | ‘3.19.0-47-generic’ | ‘#53~14.04.1-Ubuntu SMP Mon Jan 18 16:09:14 UTC 2016’ | ‘x86_64’ | ‘x86_64’  | foo@bar.com | ‘Tue Mar 8 00:52:35 2016’ | 0
```

Comment:
>  it's not possible to receive Ubuntu 14.04 from platform output, is it?

This information is already available in the version string (`14.04.1-Ubuntu`) and can be retrieved. If not, then one can also obtain information from `lsb_release` output after detecting a `posix` system. Although, that would be redundant.

> The most sense would make if system was one of [..] what system looks like on other platforms? 

The three systems would be `Linux`, `Windows` and `Darwin` based on the output that I have seen across them.

> you don't have "distribution" here

I think that the distribution information is populated only by Linux systems. So, I preferred to avoid information that might be available for a single type of system. Nevertheless, we can collect this info.

A detailed output of the information `platform` module can return across three systems in given on [PYMOTW-platform](https://pymotw.com/2/platform/) page.